### ISSUE TITLE
Inference giving 45 characters output.

### ISSUE THREAD

**Issue by AnishTadepalli** [GabrielDornelles/TorchNN-OCR#9]
Hi, I am training the model on captcha images, the model is getting upto 99.4% accuracy, and I can see the ground truth and prediction characters are same in length. But when I am using inference, I am getting 45 characters list as output. I have tried lot of things but still same type of output.
Predictions: ['g', '8', '8', '8', '8', '8', '8', '8', 'n', 'f', 'w', 'w', '7', 'e', 'e', '8', 'w', 'x', 'w', 'x', 'w', 'b', 'f', 'f', 'f', 'w', '3', 'p', 'p', 'p', '3', '3', '3', '3', '3', '3', '3', '3', '3', '3', 'p', 'p', 'p', 'p', '7']
and the characters in the image are: 2b827
This is my config:
defaults:
  - override hydra/job_logging: custom
  
processing:
  device: cpu
  image_width: 180
  image_height: 50 
  
training:
  lr: 3e-4
  batch_size: 8
  num_workers: 4
  num_epochs: 100

bools:
  DISPLAY_ONLY_WRONG_PREDICTIONS: true
  VIEW_INFERENCE_WHILE_TRAINING: true
  SAVE_CHECKPOINTS: true

paths:
  dataset_dir: ./dataset
  save_model_as: ./logs/crnn.pth

model:
  use_attention: true 
  use_ctc: true
  gray_scale: true
  dims: 256

**Comment by GabrielDornelles** [comment]
Can you provide me  your `inference.py` file?

**Comment by AnishTadepalli** [comment]
import torch
import albumentations
import numpy as np
from PIL import Image
from models.crnn import CRNN
from utils.model_decoders import decode_predictions, decode_padded_predictions

# I use "∅" to denote the blank token. This list is automatically generated at training,
# but I recommend that you hardcode your characters at evaluation
classes = ['2', '3', '4', '5', '6', '7', '8', 'b', 'c', 'd', 'e', 'f', 'g', 'm', 'n', 'p', 'w', 'x', 'y']


def inference(image_path, model):
    # Hardcoded resize
    image = Image.open(image_path).convert("L")  # Convert to grayscale
    image = image.resize((180, 50), resample=Image.BILINEAR)
    image = np.array(image)

    mean = [0.485]  # Adjust according to your data distribution
    std = [0.229]  # Adjust according to your data distribution
    aug = albumentations.Compose([albumentations.Normalize(mean, std, max_pixel_value=255.0, always_apply=True)])

    image = np.expand_dims(image, axis=-1)
    image = aug(image=image)["image"]
    image = np.transpose(image, (2, 0, 1)).astype(np.float32)
    image = image[None, ...]
    image = torch.from_numpy(image)

    if str(device) == "cuda":
        image = image.cuda()

    image = image.float()

    with torch.no_grad():
        preds, _ = model(image)

    if model.use_ctc:
        answer = decode_predictions(preds, classes)
    else:
        answer = decode_padded_predictions(preds, classes)

    return answer


if __name__ == "__main__":
    # Instantiate the model
    model = CRNN(dims=256, num_chars=len(classes), use_attention=True, use_ctc=True, grayscale=True)
    device = torch.device("cpu")
    model.to(device)
    # Load the state dictionary from the checkpoint

    loaded_model = torch.load("logs/checkpoint-99.04.pth")

    # If you want to load the state dictionary only
    loaded_model_state_dict = loaded_model.state_dict()
    model.load_state_dict(loaded_model_state_dict, strict=False)
    model.eval()

    filepath = "dataset/2bg48.png"
    answer = inference(filepath, model)
    print(f"text: {answer}")

**Comment by AnishTadepalli** [comment]
Yes, please fine the below attachment. 

[inference.txt](https://github.com/GabrielDornelles/pytorch-ocr/files/13722601/inference.txt)

**Comment by GabrielDornelles** [comment]
Please rewrite your classes list to the same that was trained, its printed in the terminal as the training starts. The first element in the list MUST be `"∅"`. 

I see you also changed a few things in the image processing (mean and std, grayscale convertion), so you need to ensure that you have done this at training code too.

**Comment by AnishTadepalli** [comment]
Hi, as you mentioned I have modified the training code (mean,std blank char), but still getting the same type of output.
I am adding the modified files.
[train.txt](https://github.com/GabrielDornelles/pytorch-ocr/files/13736522/train.txt)
[data_loading.txt](https://github.com/GabrielDornelles/pytorch-ocr/files/13736523/data_loading.txt)
[inference.txt](https://github.com/GabrielDornelles/pytorch-ocr/files/13736524/inference.txt)

**Comment by AnishTadepalli** [comment]
Hi, thanks for your time and responses. Actually at the end of the training I am getting the error as " load state dict Expected state_dict to be dict-like, got <class 'NoneType'>.",because I am giving the training.epochs=25,5. So if the best accuracy is achieved at epoch 18 in experiment 2, without saving the model I am continuing the training loop so the model weights are not loading into the model. I did small modification in train.py. now I am getting proper predictions.
[train.txt](https://github.com/GabrielDornelles/pytorch-ocr/files/13740172/train.txt)

**Comment by AnishTadepalli** [comment]
Break the training loop when accuracy is > <your threshold> then model weights will be load into the model properly
